### ISSUE TITLE
Regex:  Vietnam

### ISSUE THREAD

**Issue by cjyetman** [vincentarelbundock/countrycode#207]
The current regex for Vietnam causes a false positive on "Republic of Vietnam" if it is preceded by any character, including whitespace.

`^(?!south)(?!republic).*viet.?nam(?!.*south)|socialist.republic.of.viet.?nam|north.viet.?nam|viet.?nam.north`

e.g.
```r
countrycode(" Republic of Vietnam", 'country.name', 'country.name')
# [1] NA
# Warning messages:
#   1: In countrycode(" Republic of Vietnam", "country.name", "country.name") :
#   Some values were not matched unambiguously:  Republic of Vietnam
# 
# 2: In countrycode(" Republic of Vietnam", "country.name", "country.name") :
#   Some strings were matched more than once, and therefore set to <NA> in the result:  Republic of Vietnam,Vietnam,Republic of Vietnam
```

This causes failures on build when the tests for recognizing a country name within a sentence run because the sentence "The governement of Republic of Vietnam was reelected." matches "Vietnam" and "Republic of Vietnam".

e.g.

```r
countrycode("The governement of Republic of Vietnam was reelected.", 'country.name', 'country.name')
# [1] NA
# Warning messages:
#   1: In countrycode("The governement of Republic of Vietnam was reelected.",  :
#                       Some values were not matched unambiguously: The governement of Republic of Vietnam was reelected.
#                     
#                     2: In countrycode("The governement of Republic of Vietnam was reelected.",  :
#                                         Some strings were matched more than once, and therefore set to <NA> in the result: The governement of Republic of Vietnam was reelected.,Vietnam,Republic of Vietnam
```

**Comment by cjyetman** [comment]
FYI... the fix from #232 should makes `" Republic of Vietnam"` work, but not `"The governement of Republic of Vietnam was reelected."`. Reopening so that we don't miss that.

``` r
countrycode::countrycode("The governement of Republic of Vietnam was reelected.", 'country.name', 'country.name')
#> Warning in countrycode::countrycode("The governement of Republic of Vietnam was reelected.", : Some values were not matched unambiguously: The governement of Republic of Vietnam was reelected.
#> Warning in countrycode::countrycode("The governement of Republic of Vietnam was reelected.", : Some strings were matched more than once, and therefore set to <NA> in the result: The governement of Republic of Vietnam was reelected.,Republic of Vietnam,Vietnam
#> [1] NA
```

**Comment by cjyetman** [comment]
Somewhat different, but related to the underlying regex problem here... "Socialist Republic of Vietnam" does not currently match with the current CRAN nor the dev versions, though [apparently](https://en.wikipedia.org/wiki/Vietnam) that is the current official name of Vietnam. This is further complicated by my research in https://github.com/vincentarelbundock/countrycode/issues/16#issuecomment-413881188 where I found that technically "Republic of Vietnam" refers to a country that no longer exists (and is assigned a different CoW code), so technically it should **not** match.

``` r
library(countrycode)
packageVersion('countrycode')
#> [1] '1.1.3'
countrycode("Socialist Republic of Vietnam", 'country.name', 'country.name')
#> Warning in countrycode("Socialist Republic of Vietnam", "country.name", : Some values were not matched unambiguously: Socialist Republic of Vietnam
#> Warning in countrycode("Socialist Republic of Vietnam", "country.name", : Some strings were matched more than once, and therefore set to <NA> in the result: Socialist Republic of Vietnam,Vietnam,Republic of Vietnam
#> [1] NA
```

``` r
library(countrycode)
packageVersion('countrycode')
#> [1] '1.1.3.9000'
countrycode("Socialist Republic of Vietnam", 'country.name', 'country.name')
#> Warning in countrycode("Socialist Republic of Vietnam", "country.name", : Some values were not matched unambiguously: Socialist Republic of Vietnam
#> Warning in countrycode("Socialist Republic of Vietnam", "country.name", : Some strings were matched more than once, and therefore set to <NA> in the result: Socialist Republic of Vietnam,Republic of Vietnam,Vietnam
#> [1] NA
```

**Comment by vincentarelbundock** [comment]
Useful info copied from closed issue https://github.com/vincentarelbundock/countrycode/issues/16

> I'm pretty confused about how to fix this, so I'm leaving here some info I've been collecting...
> 
> The current regex in data_static.csv for Vietnam, which should be unique and only refer to one possible entity is: ^(?!.*republic).*viet.?nam|^(?=.*socialist).*viet.?nam
> 
> There are 3 entities that could be able uniquely distinguished:
> 
> South Vietnam 1954-1975
> possible names: South Vietnam, Republic of Vietnam
> CoW: RVN - 817
> 
> North Vietnam / Vietnam 1954-1975
> possible names: North Vietnam, Democratic Republic of Vietnam
> CoW: DRV - 816
> 
> Vietnam 1976-present
> possible names: Vietnam, Viet Nam, Socialist Republic of Vietnam
> CoW: DRV - 816
> 
> Since Vietnam 1976-present is the only one that currently exists, following the philosophy of only converting codes that are valid in the current year, there should be only one regex, and it should...
> 
> match any of: "Vietnam", "Viet Nam", and "Socialist Republic of Vietnam"
> 
> exclude any of: "South Vietnam", "Republic of Vietnam", "North Vietnam", and "Democratic Republic of Vietnam"
> 
> That makes for a somewhat complicated regex, and it means that the current one probably needs to be adapted.

**Comment by vincentarelbundock** [comment]
Copied from closed issue https://github.com/vincentarelbundock/countrycode/issues/207

--

The current regex for Vietnam causes a false positive on "Republic of Vietnam" if it is preceded by any character, including whitespace.

`^(?!south)(?!republic).*viet.?nam(?!.*south)|socialist.republic.of.viet.?nam|north.viet.?nam|viet.?nam.north`

e.g.
```r
countrycode(" Republic of Vietnam", 'country.name', 'country.name')
# [1] NA
# Warning messages:
#   1: In countrycode(" Republic of Vietnam", "country.name", "country.name") :
#   Some values were not matched unambiguously:  Republic of Vietnam
# 
# 2: In countrycode(" Republic of Vietnam", "country.name", "country.name") :
#   Some strings were matched more than once, and therefore set to <NA> in the result:  Republic of Vietnam,Vietnam,Republic of Vietnam
```

This causes failures on build when the tests for recognizing a country name within a sentence run because the sentence "The governement of Republic of Vietnam was reelected." matches "Vietnam" and "Republic of Vietnam".

e.g.

```r
countrycode("The governement of Republic of Vietnam was reelected.", 'country.name', 'country.name')
# [1] NA
# Warning messages:
#   1: In countrycode("The governement of Republic of Vietnam was reelected.",  :
#                       Some values were not matched unambiguously: The governement of Republic of Vietnam was reelected.
#                     
#                     2: In countrycode("The governement of Republic of Vietnam was reelected.",  :
#                                         Some strings were matched more than once, and therefore set to <NA> in the result: The governement of Republic of Vietnam was reelected.,Vietnam,Republic of Vietnam
```

**Comment by NilsEnevoldsen** [comment]
I think "recognizing a country name within a sentence" should be out of scope.

**Comment by vincentarelbundock** [comment]
Agreed.

**Comment by vincentarelbundock** [comment]
Fixed by c64acad
